src/solvers/refinement/string_refinement.cpp

```diff
@@ -0,0 +1,869 @@
+  {
+    exprt witness=generator.get_witness_of(not_contains_axioms[i], zero);
+    exprt val=get(witness);
+    violated.push_back(std::make_pair(i, val));
+  }
+
+  if(violated.empty())
+  {
+    debug() << "no violated property" << eom;
+    return true;
+  }
+  else
+  {
+    debug() << violated.size() << " string axioms can be violated" << eom;
+
+    if(use_counter_example)
+    {
+      std::vector<string_constraintt*> new_axioms(violated.size());
+
+      // Checking if the current solution satisfies the constraints
+      for(size_t i=0; i<violated.size(); i++)
+      {
+        new_axioms[i]=&universal_axioms[violated[i].first];
+
+        const exprt &val=violated[i].second;
+        const string_constraintt &axiom=universal_axioms[violated[i].first];
+
+        exprt premise(axiom.premise());
+        exprt body(axiom.body());
+        implies_exprt instance(premise, body);
+        debug() << "warning: we don't eliminate the existential quantifier"
+                << eom;
+        replace_expr(axiom.univ_var(), val, instance);
+        if(seen_instances.insert(instance).second)
+          add_lemma(instance);
+        else
+          debug() << "instance already seen" << eom;
+      }
+    }
+
+    return false;
+  }
+}
+
+/*******************************************************************\
+
+Function: string_refinementt::map_representation_of_sum
+
+  Inputs: an expression with only addition and substraction
+
+ Outputs: a map where each leaf of the input is mapped to the number of times
+          it is added. For instance, expression $x + x - y$ would give the map
+          x -> 2, y -> -1.
+
+\*******************************************************************/
+
+std::map<exprt, int> string_refinementt::map_representation_of_sum(const exprt &f)
+{
+  // number of time the leaf should be added (can be negative)
+  std::map<exprt, int> elems;
+
+  std::vector<std::pair<exprt, bool> > to_process;
+  to_process.push_back(std::make_pair(f, true));
+
+  while(!to_process.empty())
+  {
+    exprt cur=to_process.back().first;
+    bool positive=to_process.back().second;
+    to_process.pop_back();
+    if(cur.id()==ID_plus)
+    {
+      to_process.push_back(std::make_pair(cur.op1(), positive));
+      to_process.push_back(std::make_pair(cur.op0(), positive));
+    }
+    else if(cur.id()==ID_minus)
+    {
+      to_process.push_back(std::make_pair(cur.op1(), !positive));
+      to_process.push_back(std::make_pair(cur.op0(), positive));
+    }
+    else if(cur.id()==ID_unary_minus)
+    {
+      to_process.push_back(std::make_pair(cur.op0(), !positive));
+    }
+    else
+    {
+      if(positive)
+        elems[cur]=elems[cur]+1;
+      else
+        elems[cur]=elems[cur]-1;
+    }
+  }
+  return elems;
+}
+
+
+exprt string_refinementt::sum_over_map(std::map<exprt, int> & m, bool negated)
+{
+  exprt sum=refined_string_typet::index_of_int(0);
+  mp_integer constants=0;
+
+  for(auto it : m)
```

Comment:
for(const auto &it : m)  [though "it" is a questionable name as it isn't an iterator]